woocommerce/woocommerce
# add variable product support to api package

### Changes proposed in this Pull Request:

This PR add variable product and product variation support to the api package. This includes the following:

- Updates `products.variable` in the local `default.json` to include product properties for testing the api
- Adds `products.variations`  in the local `default.json` for testing the api
- Adds child model creation support
- Updated package tests for child model creation
- Adds E2E tests for variable product and product variations
- Splits out `AbstractProductData` from `AbstractProduct` where product data is the subset of properties for variations
- Splits out `AbstractAttribute` from `ProductAttribute` to support default & variation attributes
- Fixes issue with grouped products from #29214

Closes #28883 .

### How to test the changes in this Pull Request:

1. Verify CI tests in Github
```
  REST API > Variable Product
    ✓ can create a variable product (195ms)
    ✓ can add variations (503ms)
    ✓ can retrieve a transformed variable product (107ms)
    ✓ can retrieve a transformed product variations (77ms)
    ✓ can delete a variation (4ms)
    ✓ can delete a variable product (3ms)
```
2. Run E2E tests locally
3. Edit `tests/e2e/core-tests/api/variable-product.test.js` and comment out the last two tests which delete the test data
4. Edit `products.variable` and `products.variations` in `tests/e2e/config/default.json`
5. Run `npx wc-e2e test:e2e tests/e2e/specs/rest-api/variable-product.js`
6. Verify that the product was created correctly using the web browser
7. Repeat step 3 to just comment out the product delete
8. Re-run step 5 and verify using the browser that the last variation in default.json was deleted.

### Other information:


### Changelog entry

N/A
